File: packages/policy-core-ref/src/jcs.ts
// RFC 8785 JSON Canonicalization Scheme
// Produces byte-identical output for identical semantic inputs.
// SEMANTICS.md §9: JCS(x) with SHA-256 digest over label || 0x00 || JCS(normalized(x))

import type { JsonRecord } from "./types";

export function jcs(value: unknown): Uint8Array {
  const canonical = sortObjectMembers(value);
  const json = stringify(canonical);
  return new TextEncoder().encode(json);
}

export function digest(label: string, value: unknown): string {
  const hasher = new Bun.CryptoHasher("sha256");
  hasher.update(label);
  hasher.update(new Uint8Array([0])); // 0x00 separator
  hasher.update(jcs(value));
  return hasher.digest("hex");
}

// Sort all object members by key (ascending UTF-8 order).
// Recursively sort nested structures.
function sortObjectMembers(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map(sortObjectMembers);
  }
  if (isObject(value)) {
    const sorted: JsonRecord = {};
    const keys = Object.keys(value).sort();
    for (const key of keys) {
      sorted[key] = sortObjectMembers((value as JsonRecord)[key]);
    }
    return sorted;
  }
  return value;
}

// RFC 8785 canonicalization: serialize value to JSON string with canonical number format.
function stringify(value: unknown): string {
  if (value === null) {
    return "null";
  }
  if (typeof value === "boolean") {
    return value ? "true" : "false";
  }
  if (typeof value === "number") {
    return stringifyNumber(value);
  }
  if (typeof value === "string") {
    return stringifyString(value);
  }
  if (Array.isArray(value)) {
    const items = value.map(stringify);
    return `[${items.join(",")}]`;
  }
  if (isObject(value)) {
    const keys = Object.keys(value).sort();
    const pairs = keys.map(
      (key) => `${stringifyString(key)}:${stringify((value as JsonRecord)[key])}`,
    );
    return `{${pairs.join(",")}}`;
  }
  throw new TypeError(`cannot stringify ${typeof value}`);
}

// RFC 8785 requires careful handling of floating-point numbers.
// -0 and 0 are equal, and numbers must serialize in a canonical form.
function stringifyNumber(value: number): string {
  if (!Number.isFinite(value)) {
    throw new TypeError("non-finite number in JSON");
  }
  // -0 === 0 in IEEE-754 equality, but they serialize differently.
  // Canonical form: return "0" for both.
  if (Object.is(value, -0)) {
    return "0";
  }
  // For finite numbers, use JSON.stringify which produces IEEE-754 canonical form.
  return JSON.stringify(value);
}

// RFC 8785: escape all control characters and quotes/backslashes.
function stringifyString(value: string): string {
  let result = '"';
  for (let i = 0; i < value.length; i++) {
    const char = value[i];
    if (!char) return `${result}"`; // Safety check, should never occur
    const code = char.charCodeAt(0);
    switch (char) {
      case '"':
      case "\\":
        result += `\\${char}`;
        break;
      case "\b":
        result += "\\b";
        break;
      case "\f":
        result += "\\f";
        break;
      case "\n":
        result += "\\n";
        break;
      case "\r":
        result += "\\r";
        break;
      case "\t":
        result += "\\t";
        break;
      default:
        if (code < 0x20) {
          result += `\\u${code.toString(16).padStart(4, "0")}`;
        } else {
          result += char;
        }
    }
  }
  result += '"';
  return result;
}

function isObject(value: unknown): value is JsonRecord {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

File: crates/policy-core/src/jcs.rs
// RFC 8785 JSON Canonicalization Scheme
// Produces byte-identical output for identical semantic inputs.
// SEMANTICS.md §9: JCS(x) with SHA-256 digest over label || 0x00 || JCS(normalized(x))

use serde_json::Value;
use sha2::{Digest, Sha256};

/// Serialize value to canonical JSON (RFC 8785)
pub fn jcs(value: &Value) -> Vec<u8> {
    serde_jcs::to_vec(value).expect("JCS serialization failed")
}

/// Compute digest: H("label", x) = SHA256(UTF8(label) || 0x00 || JCS(normalize(x)))
pub fn digest(label: &str, value: &Value) -> String {
    let mut hasher = Sha256::new();
    hasher.update(label.as_bytes());
    hasher.update([0u8]); // 0x00 separator
    hasher.update(jcs(value));
    let hash = hasher.finalize();
    // Convert to hex string
    hash.iter()
        .map(|b| format!("{:02x}", b))
        .collect::<String>()
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_jcs_empty_object() {
        let val = json!({});
        let result = jcs(&val);
        assert_eq!(result, b"{}");
    }

    #[test]
    fn test_jcs_simple_object() {
        let val = json!({"a": 1, "b": 2});
        let result = jcs(&val);
        // JCS sorts keys
        assert_eq!(result, b"{\"a\":1,\"b\":2}");
    }

    #[test]
    fn test_jcs_number_canonical() {
        let val = json!({"x": 1.5});
        let result = jcs(&val);
        assert_eq!(result, b"{\"x\":1.5}");
    }

    #[test]
    fn test_digest_format() {
        let val = json!({"test": "value"});
        let d = digest("test-label", &val);
        // Should be hex string of SHA-256 (64 chars)
        assert_eq!(d.len(), 64);
        assert!(d.chars().all(|c| c.is_ascii_hexdigit()));
    }
}
